File: cflib_2_12/cplus/libcflib/CFLibCachedMessageLog.cpp
/*
 *  MSS Code Factory CFLib 2.12
 *
 *	Copyright (c) 2020 Mark Stephen Sobkow
 *	
 *	Licensed under the Apache License, Version 2.0 (the "License");
 *	you may not use this file except in compliance with the License.
 *	You may obtain a copy of the License at
 *	
 *	    http://www.apache.org/licenses/LICENSE-2.0
 *	
 *	Unless required by applicable law or agreed to in writing, software
 *	distributed under the License is distributed on an "AS IS" BASIS,
 *	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *	See the License for the specific language governing permissions and
 *	limitations under the License.
 */

#include <iostream>
#include <iomanip>
#include <sstream>

#include <cflib/CFLib.hpp>
#include <cflib/ICFLibMessageLog.hpp>
#include <cflib/CFLibCachedMessageLog.hpp>

namespace cflib {

	CFLibCachedMessageLog::~CFLibCachedMessageLog() {
	}

	CFLibCachedMessageLog::CFLibCachedMessageLog() {
		ndnt = 1;
	}

	void CFLibCachedMessageLog::clearCache() {
		static const std::string S_Empty( "" );
		cacheContents = S_Empty;
	}

	const std::string& CFLibCachedMessageLog::getCacheContents() const {
		return( cacheContents );
	}

	int CFLibCachedMessageLog::getMessageLogIndent() const {
		return( ndnt );
	}

	void CFLibCachedMessageLog::dedent() {
		if( ndnt > 1 ) {
			ndnt --;
		}
		else {
			ndnt = 1;
		}
	}

	void CFLibCachedMessageLog::indent() {
		ndnt ++;
	}

	void CFLibCachedMessageLog::message( const std::string& msg ) {
		int i;
		for( i = (( cacheContents.empty() ) ? 1 : 0 ); i < ndnt; i ++ ) {
			cacheContents.push_back( '\t' );
		}
		cacheContents.append( msg );
		if( cacheContents.back() != '\n' ) {
			cacheContents.push_back( '\n' );
		}
	}

	void CFLibCachedMessageLog::openLogFile( const std::string& fileName ) {
		// Do-nothing stub; cached message logs don't write to files
	}

	void CFLibCachedMessageLog::closeLogFile() {
		// Do-nothing stub; cached message logs don't write to files
	}
}

File: cflib_2_12/java/cflib/src/org.msscf.msscf.cflib.CFLib/org/msscf/msscf/cflib/CFLib/CFLibCachedMessageLog.java
/*
 *  MSS Code Factory CFLib 2.12
 *
 *	Copyright 2020 Mark Stephen Sobkow
 *
 *	This file is part of MSS Code Factory.
 *
 *	MSS Code Factory is free software: you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation, either version 3 of the License, or
 *	(at your option) any later version.
 *
 *	MSS Code Factory is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *
 *	You should have received a copy of the GNU General Public License
 *	along with MSS Code Factory.  If not, see https://www.gnu.org/licenses/.
 *
 *	Donations to support MSS Code Factory can be made at
 *	https://www.paypal.com/paypalme2/MarkSobkow
 *
 *	Please contact Mark Stephen Sobkow at msobkow@sasktel.net for commercial licensing.
 */

package org.msscf.msscf.cflib.CFLib;

import java.io.*;
import java.util.*;

public class CFLibCachedMessageLog
implements ICFLibMessageLog {

	private StringBuffer cacheContents = new StringBuffer();
	private int		   	indent = 0;
	
	/**
	 *	Default constructor.
	 */
	public CFLibCachedMessageLog() {
		indent = 0;
	}

	public void clearCache() {
		cacheContents.setLength( 0 );
	}

	public String getCacheContents() {
		return( cacheContents.toString() );
	}
	
	public int getMessageLogIndent() {
		return( indent );
	}
	
	public synchronized void dedent() {
		if( indent > 0 ) {
			indent --;
		}
		else {
			indent = 0;
		}
	}
	
	public synchronized void indent() {
		indent ++;
	}
	
	public synchronized void message( String msg ) {

		if( msg == null ) {
			return;
		}

		int			i;
		for( i = ( ( cacheContents.length() <= 0 ) ? 1 : 0 ); i < indent; i ++ ) {
			cacheContents.append( "\t" );
		}
		cacheContents.append( msg );
		if( ! msg.endsWith( "\n" ) ) {
			cacheContents.append( "\n" );
		}
	}

//	OLD API's that need to go away

	public PrintStream getPrintStream() {
		return( null );
	}

	public void openLogFile( String fileName )
		throws FileNotFoundException
	{
		// Do-nothing stub; cached message logs don't write to files
	}
	
	public void closeLogFile() {
		// Do-nothing stub; cached message logs don't write to files
	}
}
